File: source/python/simple_repl.py
inputs = [
    "2 + 3",
    "10 * 4",
    "100 / 7",
    "50 - 18",
    "2 + 3 * 4",
]

def tokenize(expr):
    tokens = []
    num = ""
    for ch in expr:
        if ch == " ":
            if num:
                tokens.append(num)
                num = ""
        elif ch in "+-*/":
            if num:
                tokens.append(num)
                num = ""
            tokens.append(ch)
        else:
            num += ch
    if num:
        tokens.append(num)
    return tokens

def evaluate(expr):
    tokens = tokenize(expr)
    if not tokens:
        return 0
    result = float(tokens[0])
    i = 1
    while i < len(tokens):
        op = tokens[i]
        operand = float(tokens[i + 1])
        if op == "+": result += operand
        elif op == "-": result -= operand
        elif op == "*": result *= operand
        elif op == "/": result /= operand
        i += 2
    return result

print("=== Simple Calculator ===")
for inp in inputs:
    result = evaluate(inp)
    int_result = int(result)
    if float(int_result) == result:
        print(f"> {inp} = {int_result}")
    else:
        print(f"> {inp} = {result}")
print("=== Done ===")

File: source/javascript/simple_repl.js
// Simple calculator REPL with predefined inputs
const inputs = ["2 + 3", "10 * 4", "100 / 7", "50 - 18", "2 + 3 * 4"];

function tokenize(expr) {
    const tokens = [];
    let num = "";
    for (const ch of expr) {
        if (ch === ' ') {
            if (num !== "") { tokens.push(num); num = ""; }
        } else if ("+-*/".includes(ch)) {
            if (num !== "") { tokens.push(num); num = ""; }
            tokens.push(ch);
        } else {
            num += ch;
        }
    }
    if (num !== "") tokens.push(num);
    return tokens;
}

function evaluate(expr) {
    const tokens = tokenize(expr);
    if (tokens.length === 0) return 0;
    let result = parseFloat(tokens[0]);
    let i = 1;
    while (i < tokens.length) {
        const op = tokens[i], operand = parseFloat(tokens[i + 1]);
        if (op === "+") result += operand;
        else if (op === "-") result -= operand;
        else if (op === "*") result *= operand;
        else if (op === "/") result /= operand;
        i += 2;
    }
    return result;
}

console.log("=== Simple Calculator ===");
for (const input of inputs) {
    const result = evaluate(input);
    const intResult = Math.trunc(result);
    if (intResult === result) console.log(`> ${input} = ${intResult}`);
    else console.log(`> ${input} = ${result}`);
}
console.log("=== Done ===");
